amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -458,6 +466,90 @@ void runProcessLoop() {
+            } else {
+                final Set<String> processedChildShardLeases = new HashSet<>();
+
+                for (String childShard : lease.childShardIds()) {
+                    final String childShardLeaseKey = ShardInfo.getLeaseKey(completedShard, childShard);
+                    final Lease childShardLease = Optional.ofNullable(leaseCoordinator.leaseRefresher().getLease(childShardLeaseKey)).orElseThrow(
+                            () -> new IllegalStateException("Current lease " + childShardLeaseKey + " not found in lease table."));
+
+                    if (!childShardLease.checkpoint().equals(ExtendedSequenceNumber.TRIM_HORIZON)
+                            && !childShardLease.checkpoint().equals(ExtendedSequenceNumber.AT_TIMESTAMP)) {
+                        processedChildShardLeases.add(childShardLease.leaseKey());
+                    }
+                }
+
+                if (allParentShardLeasesDeleted(lease, completedShard) && Objects.equals(childShardKeys, processedChildShardLeases)) {
+                    log.info("Deleting lease {} as it has been completely processed and processing of child shards has begun.",
+                            lease);
+                    leaseCoordinator.leaseRefresher().deleteLease(lease);
+                    leaseCoordinator.dropLease(lease);
+                }
+            }
+        }
+    }
+
+    private void cleanupGarbageShards(Collection<ShardInfo> garbageShards) throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+        for (ShardInfo garbageShard : garbageShards) {
+            final Lease lease = leaseCoordinator.getCurrentlyHeldLease(ShardInfo.leaseKeyProvider.apply(garbageShard));
+            if (allParentShardLeasesDeleted(lease, garbageShard)) {
+                log.info("Deleting lease {} as it is not present in the stream.");
+                leaseCoordinator.leaseRefresher().deleteLease(lease);
+                leaseCoordinator.dropLease(lease);
+            }
+        }
+    }
+
+    private List<ShardInfo> getGarbageShards(List<ShardInfo> shardInfoAssignments) throws ExecutionException, InterruptedException {
+        final List<ShardInfo> garbageShards = new ArrayList<>();
+
+        for (ShardInfo shardInfo : shardInfoAssignments) {
+            final GetShardIteratorRequest request = GetShardIteratorRequest.builder()
+                    .streamName(getStreamIdentifier(shardInfo.streamIdentifierSerOpt()).streamName())
+                    .shardId(shardInfo.shardId())
+                    .shardIteratorType(ShardIteratorType.TRIM_HORIZON)
+                    .build();
+
+            try {
+                retrievalConfig().kinesisClient().getShardIterator(request).get();
+            } catch (Throwable t) {
+                if (t.getCause() instanceof ResourceNotFoundException) {
+                    garbageShards.add(shardInfo);
+                }
+            }
```

Comment:
Can we move this logic as part of ShardConsumer process task to avoid this being called for every execution of runprocess loop ?